tools/pypi_rss/main.go

```diff
@@ -0,0 +1,224 @@
+	latestEtag string
+	handled    map[string]map[string]bool
+	tracker    feed.Tracker
+	apiURL     *url.URL
+	mode       schema.ExecutionMode
+	queue      taskqueue.Queue
+}
+
+func (l *listener) trim() error {
+	// TODO: trim handled
+	return nil
+}
+
+func (l *listener) handle(ctx context.Context, events []Event) error {
+	var count int
+	toRun := []rebuild.Target{}
+	for _, e := range events {
+		t := rebuild.Target{
+			Ecosystem: rebuild.PyPI,
+			Package:   e.Pkg,
+			Version:   e.Version,
+			// TODO: Inlude the artifact from the event.
+			Artifact: "",
+		}
+		tracked, err := l.tracker.IsTracked(t)
+		if err != nil {
+			return errors.Wrap(err, "checking if tracked")
+		}
+		if !tracked {
+			continue
+		}
+		if _, ok := l.handled[e.Pkg]; !ok {
+			l.handled[e.Pkg] = make(map[string]bool)
+		}
+		if _, ok := l.handled[e.Pkg][e.Version]; ok {
+			continue
+		}
+		if slices.Contains(toRun, t) {
+			continue
+		}
+		toRun = append(toRun, t)
+		log.Printf("tracked package was updated: %s@%s at %v", e.Pkg, e.Version, e.Time)
+		count += 1
+	}
+	if len(toRun) == 0 {
+		return nil
+	}
+	switch l.mode {
+	case schema.AttestMode:
+		for _, msg := range feed.GroupForAttest(toRun, "pypi-rss") {
+			if _, err := l.queue.Add(ctx, l.apiURL.JoinPath("rebuild").String(), msg); err != nil {
+				return errors.Wrapf(err, "adding msg to queue %+v", msg)
+			}
+		}
+	case schema.SmoketestMode:
+		for _, msg := range feed.GroupForSmoketest(toRun, "pypi-rss") {
+			if _, err := l.queue.Add(ctx, l.apiURL.JoinPath("smoketest").String(), msg); err != nil {
+				return errors.Wrapf(err, "adding msg to queue %+v", msg)
+			}
+		}
+	default:
+		return errors.Errorf("invalid mode: %s", string(l.mode))
+	}
+	for _, t := range toRun {
+		l.handled[t.Package][t.Version] = true
+	}
+	return l.trim()
+}
+
+func (l *listener) fetch(client *http.Client) ([]Event, error) {
+	// TODO: Handle pagination.
+	req, err := http.NewRequest(http.MethodGet, l.url, nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "making HTTP request")
+	}
+	req.Header.Set("If-None-Match", l.latestEtag)
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "error making HTTP request")
+	}
+	defer resp.Body.Close()
+	switch resp.StatusCode {
+	case http.StatusNotModified:
+		return []Event{}, nil
+	case http.StatusOK:
+		l.latestEtag = resp.Header.Get("Etag")
+	default:
+		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
+	}
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.Wrap(err, "error reading response body")
+	}
+	var data RSS
+	err = xml.Unmarshal(body, &data)
+	if err != nil {
+		return nil, errors.Wrap(err, "parsing XML")
+	}
+	var events []Event
+	for _, item := range data.Channel.Items {
+		eventTime, err := time.Parse(time.RFC1123, item.PubDate)
+		if err != nil {
+			return nil, errors.Wrap(err, "parsing date")
+		}
+		if matches := versionURLRegex.FindStringSubmatch(item.Link); len(matches) == 3 {
```

Comment:
hmm can this be matches != nil? i also might pull the else up into a guard clause instead of if..else